packages/app/src/cli/prompts/dev.ts

```diff
@@ -17,17 +18,86 @@ export async function selectOrganizationPrompt(organizations: Organization[]): P
   return organizations.find((org) => org.id === choice.id)!
 }
 
-export async function selectAppPrompt(apps: OrganizationApp[]): Promise<OrganizationApp> {
-  const appList = apps.map((app) => ({name: app.title, value: app.apiKey}))
+export async function selectAppPrompt(apps: OrganizationApp[], orgId: string, token: string): Promise<OrganizationApp> {
+  const appsByApiKey: {[apiKey: string]: OrganizationApp} = {}
+  const addToApiKeyCache = (apps: OrganizationApp[]) => apps.forEach((app) => (appsByApiKey[app.apiKey] = app))
+  addToApiKeyCache(apps)
+
+  const toAnswer = (app: OrganizationApp) => ({name: app.title, value: app.apiKey})
+  const appList = apps.map(toAnswer)
+
+  /* allInputs serves as a LIFO stack, so we always can pop off the latest input
+   * as the most urgent search term, then fill in intermediate steps in the
+   * background.
+   * latestRequest tracks the most recent input, so we know which result should
+   * be displayed at all times.
+   */
+  const allInputs = ['']
+  let latestRequest: string
+
+  /* Set up an event loop, searching for apps using the latest user input once
+   * per second to avoid hitting rate limits.
+   * If we don't have any previously unseen input to search for, return without
+   * action.
+   */
+  const cachedResults: {[input: string]: OrganizationApp[]} = {'': apps}
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  const fetchInterval = setInterval(async () => {
+    let input: string | undefined
+    do {
+      input = allInputs.pop()
```

Comment:
Do we need to keep track of all inputs? I get that going back is going to be faster this way, but we're also going to hit the API many times, when most likely only the results returned by the last input will be used.
What if instead we performed a debounced search only when the user has stopped typing for like 500ms. This should cover the most common scenario and only perform one query.

I still like the idea of caching results so that if you search for something, add some characters and go back we don't fetch again, but I feel like doing extra queries just in case might be a bit wasteful here.